### PR TITLE
[RW-4159][risk=no] Terms of Service UI and cleanup of create-account flow.

### PR DESCRIPTION
**Pull request by gjuggler**
The main goal of this PR was to create a Terms of Service acknowledgement UI based on mocks from the linked Jira ticket.

(During review, I'll keep this branch pushed to my [gjordan App Engine version](https://gjordan-dot-all-of-us-workbench-test.appspot.com/login).)

The resulting UI looks like this (click "Create Account", then enter invitation key, then you land on this page):

Screen Shot 2020-02-03 at 12 41 44 PM

and the "Next" button is enabled once they've read & clicked the checkboxes:

Screen Shot 2020-02-03 at 12 41 55 PM

I couldn't help but dig into a bit of clean-up on the overall structure of the sign-in component & the various sub-components it renders. I did a few things to try and clarify the flow of data / control through the components, and improve our ability to adapt the set of sign-up forms over time. (This is partly in anticipation of launch: we're going to add a new "institutional affiliation" step, and remove the "invitation key check" step.)

The clean-up work significantly blew up the size of this PR – sorry about that! In return, I added a bunch of tests.

---
**PR checklist**

- [X] This PR meets the Acceptance Criteria in the JIRA story
- [X] The JIRA story has been moved to Dev Review
- [X] This PR includes appropriate unit tests
- [X] I have run and tested this change locally

### CODE REVIEW

**Comment by gjuggler** [comment]
Hey @blrubenstein and @NehaBroad I'd like to get both of your pairs of eyes on this PR, since it affects code you've recently (or are currently) involved in.

I opted not to do a "surgical insertion" of the TOS component here, since I felt there was some low-hanging fruit for cleanup.

Recommended flow for reviewing this PR:

- Visit https://gjordan-dot-all-of-us-workbench-test.appspot.com/login and go through the create account flow to get a feel for the UI.
- Review create-account-tos.tsx and its test, which focuses on the new component.
- Review sign-in.tsx and its test, which contains the bulk of cleanups. The main thing I was trying to do here was add more type-safety around how we progress through each sub-component of the sign-up flow, and make it easier for us to change this over the next few weeks as we head towards launch.
- Review account-creation-survey.tsx and account-creation.tsx, which have a handful of changes largely as side-effects of the cleanups to sign-in.tsx.
- Ping me if you have any questions or would like to sit down and talk through any of the changes!

File: ui/src/app/pages/login/sign-in.tsx [hunk 1]
```diff
@@ -71,99 +60,188 @@ const styles = {
...
-               style={styles.template(this.props.windowSize, backgroundImages)}>
-            <img style={{height: '1.75rem', marginLeft: '1rem', marginTop: '1rem'}}
-                 src={headerImg}/>
-            <div style={{flex: `0 0 ${maxWidth}`,
-              maxWidth: maxWidth, minWidth: '25rem'}}>
-              {this.nextDirective(stepName)}
-            </div>
-          </div>
+      const backgroundImages = StepToImageConfig[this.state.currentStep];
+      return <FlexColumn style={styles.signInContainer} data-test-id='sign-in-container'>
```
Comment (gjuggler) [inline ui/src/app/pages/login/sign-in.tsx]: Dev note: it was quite a challenge to get the sign-in and Terms of Service components to be styled correctly to allow a "flexible" PDF panel with a sticky footer at the bottom to contain the checkboxes & next button. Ultimately, I took an approach of using nested FlexColumn components all the way down... this seemed to be minimally disruptive to the existing styling, though I haven't extensively tested this on very differently-sized screens. A bit of tire-kicking against my staged App Engine version would be very helpful here.

File: ui/src/app/pages/login/account-creation/account-creation-tos.tsx [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+        numPages: 0,
+      };
+    }
+
+    /**
+     * Handles the onRenderSuccess callback from the Page element at the end of the document.
+     * This sets up the intersection listener which will change state when the user scrolls to the
+     * end of the document.
+     */
+    handleLastPageRender() {
```
Comment (calbach) [inline ui/src/app/pages/login/account-creation/account-creation-tos.tsx]: private? (also for scrollToBottom?)

File: ui/src/app/pages/login/account-creation/account-creation-tos.tsx [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+              Array.from(
+                new Array(numPages),
+                (el, index) => (
+                  <Page
+                    renderAnnotationLayer={false}
+                    renderTextLayer={false}
+                    loading=''
+                    className='terms-of-service__page'
+                    width={Math.max(500, this.props.windowSize.width * .75)}
+                    key={`page_${index + 1}`}
```
Comment (calbach) [inline ui/src/app/pages/login/account-creation/account-creation-tos.tsx]: Thank you for adding this, I can't take any more log spam

File: ui/src/app/pages/login/account-creation/account-creation-tos.tsx [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+          >
+            {
+              Array.from(
+                new Array(numPages),
+                (el, index) => (
+                  <Page
+                    renderAnnotationLayer={false}
+                    renderTextLayer={false}
+                    loading=''
+                    className='terms-of-service__page'
```
Comment (calbach) [inline ui/src/app/pages/login/account-creation/account-creation-tos.tsx]: Is the __ a convention we've used elsewhere?

File: ui/src/styles.css [hunk 5]
```diff
@@ -196,3 +196,19 @@ iframe#launcher {
...
+
+/*
+ * Hides the unnecessary "select all" header from the multiselect drop-down for the
+ * Degrees multi-select in the Account Creation component.
+ */
+.create-account__degree-select .p-multiselect-header {
+  display: none;
+}
+
+.terms-of-service__page {
```
Comment (calbach) [inline ui/src/styles.css]: Our usual pattern for React components is to use inline style (despite my initial grumbling). Is there a need for this to be done in CSS? The classes here don't seem particularly general

File: ui/tsconfig.json [hunk 6]
```diff
@@ -20,7 +20,7 @@
       "dom"
     ],
     "module": "es2015",
-    "baseUrl": "./",
+    "baseUrl": "./src",
```
Comment (calbach) [inline ui/tsconfig.json]: Why this change?

File: ui/src/app/pages/login/account-creation/account-creation.tsx [hunk 7]
```diff
@@ -175,45 +147,59 @@ export const MultiSelectWithLabel = (props) => {
...
+  nonAcademicAffiliationRole: string;
+  nonAcademicAffiliationOther: string;
+}
+
 export class AccountCreation extends React.Component<AccountCreationProps, AccountCreationState> {
   private usernameCheckTimeout: NodeJS.Timer;
 
   constructor(props: AccountCreationProps) {
+    if (props.profile.institutionalAffiliations.length !== 1) {
+      throw new Error('Profile must be pre-allocated with 1 institutional affiliation.');
```
Comment (calbach) [inline ui/src/app/pages/login/account-creation/account-creation.tsx]: I don't understand this precondition. Is this for the new account creation logic, or the old?

File: ui/src/app/pages/login/account-creation/account-creation.tsx [hunk 8]
```diff
@@ -362,10 +336,14 @@ export class AccountCreation extends React.Component<AccountCreationProps, Accou
   }
 
   clearInstitutionAffiliation() {
-    this.updateInstitutionAffiliation('nonAcademicAffiliation', '');
-    this.updateInstitutionAffiliation('role', '');
-    this.updateInstitutionAffiliation('institution', '');
-    this.updateInstitutionAffiliation('other', '');
+    this.setState(currentState => {
```
Comment (calbach) [inline ui/src/app/pages/login/account-creation/account-creation.tsx]: nit: In general I think it is best practice to avoid directly mutating the innards of the state object, though this is probably a pretty innocuous case since it's happening in the state callback. I think the preferred approach here would be either lodash `set` or via shallow copy:
```
return {
   ...currentState.profile.institutionalAffiliations[0],
  role: '',
  institution: '',
  other: ''
}
```

File: ui/src/app/pages/login/account-creation/account-creation-tos.spec.tsx [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+let props: AccountCreationTosProps;
+const onCompleteSpy = jest.fn();
+
+beforeEach(() => {
+  jest.clearAllMocks();
+  props = {
+    windowSize: {width: 1700, height: 0},
+    onComplete: onCompleteSpy,
+    pdfPath: '/assets/documents/terms-of-service.pdf'
```
Comment (calbach) [inline ui/src/app/pages/login/account-creation/account-creation-tos.spec.tsx]: Presumably this last part wasn't meant to be included in the TOS content itself, as of now it's super confusing since we'd show multiple sets of checkboxes, the ones in the PDF of course are not interactive: 
image

File: ui/src/app/pages/login/account-creation/account-creation-tos.tsx [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+                    onClick={() => this.props.onComplete()}>
+              Next
+            </Button>
+          </FlexColumn>
+        </FlexRow>
+      </FlexColumn>;
+    }
+  });
+
+export default AccountCreationTos;
```
Comment (calbach) [inline ui/src/app/pages/login/account-creation/account-creation-tos.tsx]: Should the TOS component itself live somewhere outside of the account creation flow? I thought we were planning to eventually have some kind of client-side interceptor to redirect users who hadn't submitted the TOS to do so (i.e. for handling future potential TOS updates). If that work is not planned for now, probably fine to leave as-is.

File: ui/src/app/pages/login/sign-in.tsx [hunk 11]
```diff
@@ -71,99 +60,188 @@ const styles = {
...
+}
 
+interface BackgroundImageConfig {
+  backgroundImgSrc: string;
+  smallerBackgroundImgSrc: string;
+}
 
+export const StepToImageConfig = {
+  [SignInStep.LANDING]: {
+    backgroundImgSrc: '/assets/images/login-group.png',
```
Comment (calbach) [inline ui/src/app/pages/login/sign-in.tsx]: There used to be images on the account creation page - was removal intentional?

File: ui/src/app/pages/login/sign-in.tsx (hunk 1 above)
Comment (calbach) [reply ui/src/app/pages/login/sign-in.tsx]: I couldn't break it - looks pretty nice to me.

File: ui/src/app/pages/login/sign-in.tsx [hunk 12]
```diff
@@ -71,99 +60,188 @@ const styles = {
...
-              {this.nextDirective(stepName)}
-            </div>
-          </div>
+      const backgroundImages = StepToImageConfig[this.state.currentStep];
+      return <FlexColumn style={styles.signInContainer} data-test-id='sign-in-container'>
+        <FlexColumn data-test-id='sign-in-page'
+             style={styles.template(this.props.windowSize, backgroundImages)}>
+          <div><img style={{height: '1.75rem', marginLeft: '1rem', marginTop: '1rem'}}
+                    src={HEADER_IMAGE}/></div>
+          {this.renderSignInStep(this.state.currentStep)}
```
Comment (calbach) [inline ui/src/app/pages/login/sign-in.tsx]: One thing that looks a bit wrong is the lack of any margin under the "submit" button on the demographics survey. That might be an existing issue, so feel free to ignore it here. Note, this screenshot shows the bottom of the window, fully scrolled:

image

File: ui/src/app/pages/login/sign-in.tsx (hunk 11 above)
Comment (calbach) [reply ui/src/app/pages/login/sign-in.tsx]: Nevermind. Existing behavior and it matches the updated mocks.

File: ui/src/app/pages/login/sign-in.tsx [hunk 13]
```diff
@@ -1,169 +1,248 @@
...
+        address: {
+          streetAddress1: '',
+          streetAddress2: '',
+          city: '',
+          state: '',
+          country: '',
+          zipCode: '',
+        },
+        institutionalAffiliations: [
+          // We only allow entering a single institutional affiliation from the creat account
```
Comment (calbach) [inline ui/src/app/pages/login/sign-in.tsx]: s/creat/create/

File: ui/src/app/pages/login/account-creation/account-creation.tsx (hunk 7 above)
Comment (calbach) [reply ui/src/app/pages/login/account-creation/account-creation.tsx]: Thanks, this seems fine. I think the only thing that may be confusing here is that typically this profile is going to be passed in an empty state. My first thought on reading this is that perhaps we're expecting the affiliation to already be filled in. A comment here explaining that it will typically be just a placeholder empty affiliation might make this more clear.

File: ui/src/app/pages/login/account-creation/account-creation.tsx (hunk 8 above)
Comment (calbach) [reply ui/src/app/pages/login/account-creation/account-creation.tsx]: My typical flow is:
- Read normal lodash docs
- Google lodash fp playground and poke at it until I can confirm understanding

Sometimes the mapping from lodash to a side-effect free functional interface is straightforward, other times it's very non-obvious.

Also there are some docs here, but I don't remember the quality:
https://gist.github.com/jfmengels/6b973b69c491375117dc
